projection.go

```diff
@@ -3,41 +3,72 @@ package dogma
+	// HandleEvent updates the projection to reflect the occurrence of event
+	// message m.
+	//
+	// If nil is returned, the projection has been updated successfully.
+	//
+	// If an error is returned, the projection SHOULD be left in the state it
+	// was before HandleEvent() was called.
+	//
+	// The engine SHOULD provide "at-least-once" delivery gaurantees to the
+	// handler. That is, the engine should call HandleEvent() with the same
+	// event message until it a nil error is returned.
+	//
+	// The engine MUST NOT call HandleEvent() with any message of a type that
+	// has not been routed to this handler by a prior call to Configure(). If
+	// any such message is passed, the implementation MUST panic with the
+	// UnexpectedMessage value.
 	//
-	// s provides access to the operations available within the scope of handling
-	// m.
+	// The engine MAY provide gaurantees about the order in which event messages
+	// will be passed to HandleEvent(), however in the interest of engine
+	// portability the implementation SHOULD NOT assume that HandleEvent() will
+	// be called with events in the same order that they were recorded.
 	//
-	// It panics with the UnexpectedMessage value if m is not one of the event
-	// types that is routed to this handler via Configure().
+	// The engine MAY call HandleEvent() from multiple goroutines concurrently.
 	HandleEvent(ctx context.Context, s ProjectionEventScope, m Message) error
 }
 
-// ProjectionConfigurer is an interface implemented by the engine and used
-// by the application to configure options related to a ProjectionMessageHandler.
+// ProjectionConfigurer is an interface implemented by the engine and used by
+// the application to configure options related to a ProjectionMessageHandler.
 //
 // It is passed to ProjectionMessageHandler.Configure(), typically upon
 // initialization of the engine.
 //
 // In the context of this interface, "the handler" refers to the handler on
 // which Configure() has been called.
 type ProjectionConfigurer interface {
-	// Name sets the name of the handler. Each handler within an application must
-	// have a unique name.
+	// Name sets the name of the handler.
+	//
+	// It MUST be called exactly once within a single call to Configure().
+	//
+	// Each handler within an application MUST have a unique, non-empty name.
 	Name(n string)
 
-	// RouteEventType configures the engine to route events of the same type as m
-	// to the handler.
+	// RouteEventType configures the engine to route event messages of the same
+	// type as m to the handler.
+	//
+	// It MUST be called at least once within a call to Configure(). It MUST NOT
+	// be called more than once with an event message of the same type.
+	//
+	// Multiple handlers within a single application MAY receive event messages
+	// of the same type.
+	//
+	// The "content" of m MUST NOT be used, inspected, or treated as meaningful
+	// in any way, only its runtime type information.
 	RouteEventType(m Message)
```

Comment:
I know I might be very late reviewing this argument, but just a thought, should we pass the reflect.Type value  as an argument here (not the Message type) to conform with the method name?

Reply:
Not too late, this is why we're in the 0.x.y semver range :)

I've deliberately avoided having any part of the public interface use `reflect.Type`, for a few reasons:

1. Asking for an implementation of `dogma.Message` is unambiguous. Even though the interface is currently empty, it is widely used throughout the API and we can expect application developers to understand its purpose.
2. If we *were* to add methods to `Message` in the future, we don't have to validate that the `reflect.Type` value we're given actually adheres to that interface - it's caught by the compiler.
3. Constructing `reflect.Type` values can be rather noisy. For example, `reflect.TypeOf((*messages.SomeMessageType)(nil))` is a lot less clear than `&messages.SomeMessageType{}`. As an aside all methods in `dogma` that accept a `dogma.Message`, but only use its type have a name ending in `Type`.

Obviously it's not perfect the way it is, but there's fewer edge cases to guard against this way (we can basically check for `nil`, anything else is a valid message).

I am of course open to changing this, but IMO this is the best API we can provide to users as the language currently stands. I you feel this warrants further discussion let's do that in a separate issue.

Reply:
That sounds quite convincing, thanks! I don't really think that any more conversation on this topic are necessary.

Reply:
It would be nice to have generics for cases like this - `RouteEvent<messages.MyMessage>()`.  Maybe one day :P